src/lib/SaferNames/NameCore.hs

```diff
@@ -0,0 +1,294 @@
+
+instance InjectableE (Name s) where
+  injectionProofE _ name = unsafeCoerceE name
+
+-- Here's the reasoning informally. Let's say `x` is the name of the binder.
+-- The scopes are related like this:
+--   l  = n  ++ [x]
+--   l' = n' ++ [x]
+-- if n' is observably fresh wrt n then l' is observably fresh wrt l. This is because every
+-- exposed name in l' is either
+--   (1) x itself, which is exposed in both l and l', thus l' doesn't shadow it, or
+--   (2) a name other than x exposed by n', which we know is observably fresh
+--       wrt n and thus also observably fresh wrt n++[x]
+instance InjectableB (NameBinder s) where
+  injectionProofB  _ b cont = cont UnsafeMakeObservablyFresh $ unsafeCoerceB b
+  boundNames (UnsafeMakeBinder name) = UnsafeMakeNameSet (S.singleton name)
+
+instance (forall s. InjectableE s => InjectableE (v s)) => InjectableE (NameMap v i) where
+
+-- === environments ===
+
+-- The `NameMap` type is purely an optimization. We could do everything using
+-- the safe API by defining:
+--    type NameMap v i o = (NameSet i, forall s. Name s i -> v s o)
+-- Instead, we use this unsafely-implemented data type for efficiency, to avoid
+-- long chains of case analyses as we extend environments one name at a time.
+
+data NameMap
+  (v::E -> E)  -- env payload, as a function of the static data type
+  (i::S)       -- scope parameter for names we can look up in this env
+  (o::S)       -- scope parameter for the values stored in the env
+  = UnsafeMakeNameMap
+      (M.Map RawName (EnvVal v o))
+      (S.Set RawName)  -- cached name set as an optimization, to avoid the O(n)
+                       -- map-to-set conversion
+
+lookupNameMap :: NameMap v i o -> Name s i -> v s o
+lookupNameMap (UnsafeMakeNameMap m _) name@(UnsafeMakeName rawName) =
+  case M.lookup rawName m of
+    Nothing -> error "Env lookup failed (this should never happen)"
+    Just d -> fromEnvVal name d
+
+emptyNameMap :: NameMap v (i:=>:i) o
+emptyNameMap = UnsafeMakeNameMap mempty mempty
+
+singletonNameMap :: NameBinder s i i' -> v s o -> NameMap v (i:=>:i') o
+singletonNameMap b x = case b of
+  Ignore -> emptyNameMap
+  UnsafeMakeBinder name ->
+    UnsafeMakeNameMap (M.singleton name $ toEnvVal x) (S.singleton name)
+
+concatNameMaps :: NameMap v (i1:=>:i2) o
+               -> NameMap v (i2:=>:i3) o
+               -> NameMap v (i1:=>:i3) o
+concatNameMaps (UnsafeMakeNameMap m1 s1) (UnsafeMakeNameMap m2 s2) =
+  UnsafeMakeNameMap (m2 <> m1) (s2 <> s1)  -- flipped because Data.Map uses a left-biased `<>`
+
+extendNameMap :: NameMap v i o -> NameMap v (i:=>:i') o -> NameMap v i' o
+extendNameMap (UnsafeMakeNameMap m1 s1) (UnsafeMakeNameMap m2 s2) =
+  UnsafeMakeNameMap (m2 <> m1) (s2 <> s1)
+
+nameMapNames :: NameMap v i o -> NameSet i
+nameMapNames (UnsafeMakeNameMap _ s) = UnsafeMakeNameSet s
+
+-- === handling the dynamic/heterogeneous stuff for Env ===
+
+data EnvVal (v::E->E) (n::S) where
+  EnvVal :: TypeRep s -> v s n -> EnvVal v n
```

Comment:
Ha, you're right that `n::S` and `e::E` are reasonable enough but `v :: E -> E` starts to get a bit silly. There are only two instantiations of `v:: E -> E` so far, and they might clarify things. There's the identity map, which is used by `Scope` in `Name.hs`. It just says that if you have a `Name s n` you can query the scope to get a (newtype-wrapper) `s n`. Concretely, in the core IR we have `Name TypedBindeInfo n` for ordinary let/lambda-bound names, and `Name DataDef n` for data definitions. But I imagine we might want to start having names for modules, for pointers, for transformation specializations and so on. Then there's `SubstVal` which plays a GADT trick to say that you have two cases -- either the `s` parameter of the name you're looking up matches a particular type, in which case you get, say, an `Atom`, or else it doesn't, in which case you merely get a new name.

Reply:
Thanks, that's helpful. Does the `E -> E` pattern appear in lots of other places, or is it just in `EnvVal`? I wonder if it would be worth having an alias for it with documentation saying what some examples look like? Or just have more examples listed here?